### DISCUSSION TITLE
Better ways to bake VoxelGI

### DISCUSSION THREAD

**Discussion by KrisCatboy** [godotengine/godot-proposals#6352] [3D]
Currently VoxelGI is baked on the CPU, hardcoded to rely on `BaseMaterial3D`'s properties. Even then, its results are crude and inaccurate. And it has to give up completely on custom `ShaderMaterials`, falling back to a default grey material ....

Ideally, this is done on GPU, or in some way which runs the actual `Shader`'s vertex and fragment codeblocks on each voxel to get accurate values and allow custom shaders to be baked.

I've thought of a number of vague solutions on how it could be done:
- directly porting the voxelizer to a GPU compute shader, which is compiled with each `Shader`'s vertex() and fragment() codeblocks (this is what I initially came up with, but it seems too overblown... I can go into further detail if needed)
- keeping the current voxelizer and just running the shader code on CPU instead (not sure how feasible this is -- is GLSL difficult to write a parser for?)
- rendering to slices of the 3D albedo and emission textures with orthogonal projection and very short near-farz range (this one seems to me like the most feasible one, requiring the least work)

All of these in effect act as faux scene shaders for the purpose of being able to use the actual `Shader`'s output albedo and emission, making VoxelGI lighting and reflections appear accurate.
There'll have to be special handling for view-dependent built-ins (`VIEW`, camera matrices, etc.), which I'm not sure how to address best.
It'd also be ideal to let shaders detect being part of a VoxelGI bake, so that they can adjust their output to have a desired baked representation.

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot/issues/64124.

VoxelGI is planned to get a rework in the long run to better handle this kind of issue (in a future 4.x release).

**Reply by KrisCatboy** [reply]
I'm interested to know what kind of rework exactly is planned, if there is an answer.

**Reply by Calinou** [reply]
It's not set in stone yet, but this is something reduz plans to do. It will likely involve using the GPU for baking static VoxelGI (similar to how dynamic VoxelGI is rendered).

**Comment by clayjohn** [comment]
We plan on eventually replacing the VoxelGI voxelizer with the one used by SDFGI, then bakes can mostly take place on the GPU and should be a bit faster. The code for it is all there (either used by SDFGI or by the VoxelGI dynamic updates) and just needs to be reconfigured.
